### ISSUE TITLE
[bug]: Asterisk lockup when using realtime dialplan or any other non-core pbx engine

### ISSUE THREAD

**Issue by creslin287** [asterisk/asterisk#668]
### Severity

Major

### Versions

18.22.0, 20.7.0, 21.2.0

### Components/Modules

pbx.c

### Operating Environment

Requires use of realtime dialplan.  Was discovered with a dialplan that used an include that contained a switch to a realtime backend.

### Frequency of Occurrence

Occasional

### Issue Description

A complete lockup can occur when using a realtime dialplan backend.  When a realtime dialplan does a lookup to an included context with an external pbx switch (and possible in general when referencing a any realtime backed context) at the same time as a CNG tone is detected on a channel within an autoservice thread, Asterisk can completely lock up.  I just got done going through a core dump with 400 threads ultimately blocked due to not being able to get access to conlock in main/pbx.c.

Here are the two most relevant thread stacks from a core dump of a lockup that demonstrate the issue:

> Thread 64 (Thread 0x7f523e011640 (LWP 931102)):
#0  futex_wait (private=0, expected=2, futex_word=0x55fc844c5f20 <conlock>) at ../sysdeps/nptl/futex-internal.h:146
#1  __GI___lll_lock_wait (futex=futex@entry=0x55fc844c5f20 <conlock>, private=0) at ./nptl/lowlevellock.c:49
#2  0x00007f5303cf205d in lll_mutex_lock_optimized (mutex=0x55fc844c5f20 <conlock>) at ./nptl/pthread_mutex_lock.c:48
#3  ___pthread_mutex_lock (mutex=mutex@entry=0x55fc844c5f20 <conlock>) at ./nptl/pthread_mutex_lock.c:128
#4  0x000055fc842a26fc in __ast_pthread_mutex_lock (filename=filename@entry=0x55fc844147d4 "pbx.c", lineno=lineno@entry=8516, func=func@entry=0x55fc84417e10 <__PRETTY_FUNCTION__.22> "ast_rdlock_contexts", mutex_name=mutex_name@entry=0x55fc844148f9 "&conlock", t=t@entry=0x55fc844c5f20 <conlock>) at lock.c:347
#5  0x000055fc842c796b in ast_rdlock_contexts () at pbx.c:8516
#6  pbx_extension_helper (c=c@entry=0x7f525448ae80, context=context@entry=0x7f525448b840 "ivr_8015", exten=exten@entry=0x7f52a0ae04c5 "fax", priority=priority@entry=1, label=label@entry=0x0, callerid=callerid@entry=0x7f525428c7d0 "RT_16109925000", action=E_MATCH, found=0x0, combined_find_spawn=0, con=0x0) at pbx.c:2887
#7  0x000055fc842c95a9 in ast_exists_extension (c=c@entry=0x7f525448ae80, context=context@entry=0x7f525448b840 "ivr_8015", exten=exten@entry=0x7f52a0ae04c5 "fax", priority=priority@entry=1, callerid=callerid@entry=0x7f525428c7d0 "RT_16109925000") at pbx.c:4178
#8  0x00007f52a0ad81c1 in chan_pjsip_cng_tone_detected (f=<optimized out>, session=<optimized out>, ast=0x7f525448ae80) at chan_pjsip.c:799
#9  chan_pjsip_read_stream (ast=0x7f525448ae80) at chan_pjsip.c:935
#10 0x000055fc84245e17 in __ast_read (chan=chan@entry=0x7f525448ae80, dropaudio=dropaudio@entry=0, dropnondefault=dropnondefault@entry=1) at channel.c:3751
#11 0x000055fc84247870 in ast_read (chan=chan@entry=0x7f525448ae80) at channel.c:4294
#12 0x000055fc841fbb39 in autoservice_run (ign=ign@entry=0x0) at autoservice.c:135
#13 0x000055fc8435e636 in dummy_start (data=<optimized out>) at utils.c:1607 
#14 0x00007f5303ceeac3 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:442
#15 0x00007f5303d80850 in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81

and

> Thread 368 (Thread 0x7f50ec46d640 (LWP 1228733)):
#0  0x00007f5303d3f7f8 in __GI___clock_nanosleep (clock_id=clock_id@entry=0, flags=flags@entry=0, req=req@entry=0x7f50ec469250, rem=rem@entry=0x0) at ../sysdeps/unix/sysv/linux/clock_nanosleep.c:78
#1  0x00007f5303d44677 in __GI___nanosleep (req=req@entry=0x7f50ec469250, rem=rem@entry=0x0) at ../sysdeps/unix/sysv/linux/nanosleep.c:25
#2  0x00007f5303d75f2f in usleep (useconds=useconds@entry=1000) at ../sysdeps/posix/usleep.c:31
#3  0x000055fc841fc36a in ast_autoservice_stop (chan=chan@entry=0x7f525448ae80) at autoservice.c:314
#4  0x000055fc842c6076 in pbx_find_extension (chan=chan@entry=0x7f525448ae80, bypass=bypass@entry=0x0, q=q@entry=0x7f50ec469b30, context=0x7f529049e8c0 "dbndirs", exten=exten@entry=0x7f525448b890 "s", priority=priority@entry=37, label=<optimized out>, callerid=<optimized out>, action=<optimized out>) at pbx.c:2752
#5  0x000055fc842c6474 in pbx_find_extension (chan=chan@entry=0x7f525448ae80, bypass=bypass@entry=0x0, q=q@entry=0x7f50ec469b30, context=<optimized out>, exten=exten@entry=0x7f525448b890 "s", priority=priority@entry=37, label=<optimized out>, callerid=<optimized out>, action=<optimized out>) at pbx.c:2771
#6  0x000055fc842c6474 in pbx_find_extension (chan=chan@entry=0x7f525448ae80, bypass=bypass@entry=0x0, q=q@entry=0x7f50ec469b30, context=context@entry=0x7f525448b840 "ivr_8015", exten=exten@entry=0x7f525448b890 "s", priority=priority@entry=37, label=<optimized out>, callerid=<optimized out>, action=<optimized out>) at pbx.c:2771
#7  0x000055fc842c79bc in pbx_extension_helper (c=c@entry=0x7f525448ae80, context=0x7f525448b840 "ivr_8015", exten=exten@entry=0x7f525448b890 "s", priority=priority@entry=37, label=label@entry=0x0, callerid=callerid@entry=0x7f525428c7d0 "RT_16109925000", action=E_SPAWN, found=0x7f50ec46cbf4, combined_find_spawn=1, con=0x0) at pbx.c:2896
#8  0x000055fc842c9c19 in ast_spawn_extension (combined_find_spawn=1, found=<optimized out>, callerid=0x7f525428c7d0 "RT_16109925000", priority=37, exten=0x7f525448b890 "s", context=<optimized out>, c=0x7f525448ae80) at pbx.c:4203
#9  __ast_pbx_run (c=c@entry=0x7f525448ae80, args=args@entry=0x0) at pbx.c:4377
#10 0x000055fc842cb675 in pbx_thread (data=data@entry=0x7f525448ae80) at pbx.c:4701
#11 0x000055fc8435e636 in dummy_start (data=<optimized out>) at utils.c:1607
#12 0x00007f5303ceeac3 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:442
#13 0x00007f5303d80850 in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81

Thread 368 owns the conlock in pbx.c, grabbed by pbx_extension_helper(), which then calls pbx_find_extension() through various includes, eventually hitting one with a switch statement to a realtime backend.  Prior quering the backend, it does an ast_autoservice_start(), and after querying it does an ast_autoservice_stop().

ast_autoservice_stop() is supposed to return once the channel is completely removed from the autoservice thread.  Ordinarly this is fine, but if the channel within the autoservice thread (Thread 64) detects a fax when doing an ast_read(), it attempts to do a ast_exists_extension() to check for a "fax" extension, and this lookup attempts to grab the &conlock as well, which is held by Thread 368.  This causes the autoservice thread (Thread 64) to block until Thread 368, which is trying to run an ast_autoservice_stop() and wait for the autoservice thread to release whatever channel it's waiting to hold indefinitely.

So ordering is as follows:

Thread 368 (realtime lookup thread):
Grabs conlock -> waits for autoservice thread loop number to increment (and ast_read() to complete) but never happens because Thread 64 is stuck waiting for conlock

Thread  64 (autoservice thread):
Does ast_read() -> receives fax tone in channel pvt read handler and then does "fax" exten lookup which tries to wait on conlock and never is able to get it from Thread 368 due to busy loop waiting for autoservice thread counter incrementing.

My suggestion is to remove ast_autoservice_start() and ast_autoservice_stop() wrappers when calling to a realtime backend.  They were originally added to help when realtime backends took a long time to response, and sometimes it would cause DAHDI channel buffers to drop audio data and miss fax detection tones as reported in:

https://github.com/asterisk/asterisk/commit/21a22fa584436b4417aa44a131979a1d2fbe8e72

This commit is a bit complicated, but the autoservice_start() was added here.  In this version, the start happens prior to grabbing conlock (by ast_rdlock_contexts()) and stop() happens after conlock is released, which should be a safe ordering and avoid the problem above.  Because this would cause autoservice to start for very type of dialplan lookup, realtime or not, it was then changed to reduce the scope of the autoservice_start() and _stop() to just do it for an external pbx lookup in a later commit.  Unfortunately it changed the ordering to try to do an autoservice_stop() when conlock is still held, which causes the "deadlock" as mentioned above.

This is the patch that introduced the faulty ordering change:
https://github.com/asterisk/asterisk/commit/f52c0686054c0cde4ab92e89150099e79e5e6c5e

I believe the consequence of sometimes missing a couple of kernel DAHDI buffers due to removing both remaining sets of autoservice_start()/_stop() calls from pbx_find_extension() in pbx.c to be much less severe than potentially causing a complete lockup of Asterisk when using a non-native dialplan pbx switch like realtime, dundi, lua, iax2, or whatever when a fax comes in at just the wrong time.

### Asterisk Issue Guidelines

- [X] Yes, I have read the Asterisk Issue Guidelines

**Comment by alex2grad** [comment]
This bug/issue is very old https://issues-archive.asterisk.org/ASTERISK-21228

The deadlock happened again.
Thread 266 pbx_extension_helper (hold lock ast_rdlock_contexts) -> pbx_find_extension -> ast_autoservice_stop whaiting/sleeping for autoservice thread rebuilds its list
Thread 97 autoservice_run -> ast_read -> pbx_extension_helper -> waiting lock ast_rdlock_contexts

```
Thread 266 (Thread 0x7f421e7a6640 (LWP 1656)):
#0  0x00007f42e82ef775 in clock_nanosleep@GLIBC_2.2.5 () from /lib64/libc.so.6
#1  0x00007f42e82f45e3 in nanosleep () from /lib64/libc.so.6
#2  0x00007f42e832445b in usleep () from /lib64/libc.so.6
#3  0x00000000004710da in ast_autoservice_stop (chan=chan@entry=0x7f42d8cbcaf0) at autoservice.c:320
#4  0x00000000005201bf in pbx_find_extension (chan=chan@entry=0x7f42d8cbcaf0, bypass=bypass@entry=0x0, q=q@entry=0x7f421e79e660, context=context@entry=0x7f421e7a2898 "internalTEST", exten=exten@entry=0x7f421e7a28e8 "3", priority=priority@entry=1, label=<optimized out>, callerid=<optimized out>, action=<optimized out>) at pbx.c:2752
#5  0x0000000000521c48 in pbx_extension_helper (c=0x7f42d8cbcaf0, con=con@entry=0x0, context=context@entry=0x7f421e7a2898 "internalTEST", exten=exten@entry=0x7f421e7a28e8 "3", priority=priority@entry=1, label=label@entry=0x0, callerid=0x7f42d8188150 "TEST1", action=E_CANMATCH, found=0x0, combined_find_spawn=0) at pbx.c:2896
#6  0x000000000052263f in ast_canmatch_extension (c=<optimized out>, context=context@entry=0x7f421e7a2898 "internalTEST", exten=exten@entry=0x7f421e7a28e8 "3", priority=priority@entry=1, callerid=callerid@entry=0x7f42d8188150 "TEST1") at pbx.c:4193
#7  0x00007f42e41f6022 in valid_exit (qe=qe@entry=0x7f421e7a1840, digit=<optimized out>) at app_queue.c:4217
#8  0x00007f42e420f79d in wait_our_turn (reason=<optimized out>, ringing=<optimized out>, qe=<optimized out>) at app_queue.c:5995
#9  queue_exec (chan=0x7f42d8cbcaf0, data=<optimized out>) at app_queue.c:8820
#10 0x000000000052b46d in pbx_exec (c=c@entry=0x7f42d8cbcaf0, app=app@entry=0x158cde80, data=data@entry=0x7f421e7a3b10 "TEST_appointment,t,,,5400,,,,") at pbx_app.c:492
#11 0x0000000000521c09 in pbx_extension_helper (c=c@entry=0x7f42d8cbcaf0, con=con@entry=0x0, context=0x7f42d8cbd530 "DoQueue", exten=exten@entry=0x7f42d8cbd580 "s", priority=priority@entry=9, label=label@entry=0x0, callerid=<optimized out>, action=<optimized out>, found=0x7f421e7a5bc4, combined_find_spawn=<optimized out>) at pbx.c:2948
#12 0x00000000005226a2 in ast_spawn_extension (c=c@entry=0x7f42d8cbcaf0, context=<optimized out>, exten=exten@entry=0x7f42d8cbd580 "s", priority=priority@entry=9, callerid=callerid@entry=0x7f42d8188150 "TEST1", found=found@entry=0x7f421e7a5bc4, combined_find_spawn=1) at pbx.c:4203
#13 0x0000000000522bf9 in __ast_pbx_run (c=c@entry=0x7f42d8cbcaf0, args=args@entry=0x0) at pbx.c:4377
#14 0x0000000000523ff1 in pbx_thread (data=data@entry=0x7f42d8cbcaf0) at pbx.c:4701
#15 0x00000000005a477d in dummy_start (data=<optimized out>) at utils.c:1607
#16 0x00007f42e8c74e45 in start_thread () from /lib64/libpthread.so.0
#17 0x00007f42e832c4af in clone () from /lib64/libc.so.6

Thread 97 (Thread 0x7f421f663640 (LWP 5445)):
#0  0x00007f42e8c7e28b in __lll_lock_wait () from /lib64/libpthread.so.0
#1  0x00007f42e8c77040 in pthread_mutex_lock () from /lib64/libpthread.so.0
#2  0x0000000000500f98 in __ast_pthread_mutex_lock (filename=<optimized out>, lineno=<optimized out>, func=<optimized out>, mutex_name=<optimized out>, t=<optimized out>) at lock.c:353
#3  0x000000000051a6b6 in ast_rdlock_contexts () at pbx.c:8516
#4  0x000000000052196c in pbx_extension_helper (c=c@entry=0x7f42098e99f0, con=con@entry=0x0, context=0x7f42098ea430 "intercept_connectedline", exten=exten@entry=0x7f42098ea480 "s", priority=priority@entry=3, label=label@entry=0x0, callerid=0x7f4220112c30 "TEST2", action=E_SPAWN, found=0x7f421f65cb1c, combined_find_spawn=1) at pbx.c:2887
#5  0x00000000005226a2 in ast_spawn_extension (c=c@entry=0x7f42098e99f0, context=<optimized out>, exten=exten@entry=0x7f42098ea480 "s", priority=priority@entry=3, callerid=callerid@entry=0x7f4220112c30 "TEST2", found=found@entry=0x7f421f65cb1c, combined_find_spawn=1) at pbx.c:4203
#6  0x00007f42e42c1b8e in gosub_run (chan=0x7f42098e99f0, sub_args=0x7f421f65cbc0 "intercept_connectedline,s,1", ignore_hangup=0) at app_stack.c:1054
#7  0x0000000000451836 in ast_app_exec_sub (autoservice_chan=autoservice_chan@entry=0x0, sub_chan=sub_chan@entry=0x7f42098e99f0, sub_args=sub_args@entry=0x7f421f65cbc0 "intercept_connectedline,s,1", ignore_hangup=ignore_hangup@entry=0) at app.c:384
#8  0x00000000004519b7 in ast_app_run_sub (autoservice_chan=autoservice_chan@entry=0x0, sub_chan=sub_chan@entry=0x7f42098e99f0, sub_location=sub_location@entry=0x7f421f65cbc0 "intercept_connectedline,s,1", sub_args=sub_args@entry=0x7f421f65cbb0 "", ignore_hangup=ignore_hangup@entry=0) at app.c:405
#9  0x00000000004ab1db in ast_channel_connected_line_sub (autoservice_chan=autoservice_chan@entry=0x0, sub_chan=sub_chan@entry=0x7f42098e99f0, connected_info=connected_info@entry=0x7f421f65cd50, is_frame=is_frame@entry=0) at channel.c:10583
#10 0x00000000004b0200 in __ast_read (chan=chan@entry=0x7f42098e99f0, dropaudio=dropaudio@entry=0, dropnondefault=dropnondefault@entry=1) at channel.c:3870
#11 0x00000000004b064c in ast_read (chan=chan@entry=0x7f42098e99f0) at channel.c:4320
#12 0x00000000004708a1 in autoservice_run (ign=ign@entry=0x0) at autoservice.c:135
#13 0x00000000005a477d in dummy_start (data=<optimized out>) at utils.c:1607
#14 0x00007f42e8c74e45 in start_thread () from /lib64/libpthread.so.0
#15 0x00007f42e832c4af in clone () from /lib64/libc.so.6

```

**Comment by alex2grad** [comment]
It's wrong that autoservice thread  can call pbx_extension_helper.

My idea: if ast_read is running in autoservice thread and frame can be deferred (see function ast_is_deferrable_frame) or only AST_FRAME_CONTROL, then ast_read should not proceed the frame, but just return it. The autoservice thread adds this frame into deferred_frames and then autoservice_stop will proceed deferred_frames.

@jcolp, what do you think?

**Comment by jcolp** [comment]
Possibly? I don't recall autoservice details, and more specific things could come up during implementation of what you have said.
